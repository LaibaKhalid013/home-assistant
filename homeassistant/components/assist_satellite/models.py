"""Models for assist satellite."""

from enum import IntFlag, StrEnum


class AssistSatelliteState(StrEnum):
    """Valid states of an Assist satellite entity."""

    LISTENING_WAKE_WORD = "listening_wake_word"
    """Device is streaming audio for wake word detection to Home Assistant."""

    LISTENING_COMMAND = "listening_command"
    """Device is streaming audio with the voice command to Home Assistant."""

    PROCESSING = "processing"
    """Home Assistant is processing the voice command."""

    RESPONDING = "responding"
    """Device is speaking the response."""


class AssistSatelliteEntityFeature(IntFlag):
    """Supported features of Assist satellite entity."""

    ANNOUNCE = 1
    """Device supports remotely triggered announcements."""
