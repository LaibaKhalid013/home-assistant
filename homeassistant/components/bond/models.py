"""The bond integration models."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from bond_async import BPUPSubscriptions

    from .utils import BondHub


@dataclass
class BondData:
    """Data for the bond integration."""

    hub: BondHub
    bpup_subs: BPUPSubscriptions
