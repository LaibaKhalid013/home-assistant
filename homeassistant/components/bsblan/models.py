"""Models for the BSB-Lan integration."""

from dataclasses import dataclass

from bsblan.models import State


@dataclass
class BSBLanCoordinatorData:
    """BSBLan data stored in the Home Assistant data object."""

    state: State
