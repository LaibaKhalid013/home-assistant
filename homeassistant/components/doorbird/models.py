"""The doorbird integration models."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from homeassistant.config_entries import ConfigEntry

if TYPE_CHECKING:
    from .device import ConfiguredDoorBird

type DoorBirdConfigEntry = ConfigEntry[DoorBirdData]


@dataclass
class DoorBirdData:
    """Data for the doorbird integration."""

    door_station: ConfiguredDoorBird
    door_station_info: dict[str, Any]

    #
    # This integration uses a different event for
    # each entity id. It would be a major breaking
    # change to change this to a single event at this
    # point.
    #
    # Do not copy this pattern in the future
    # for any new integrations.
    #
    event_entity_ids: dict[str, str]
