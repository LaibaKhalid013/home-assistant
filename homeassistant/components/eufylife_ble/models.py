"""Models for the EufyLife integration."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from eufylife_ble_client import EufyLifeBLEDevice


@dataclass
class EufyLifeData:
    """Data for the EufyLife integration."""

    address: str
    model: str
    client: EufyLifeBLEDevice
