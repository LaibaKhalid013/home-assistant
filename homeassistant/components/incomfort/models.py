"""Models for Intergas InComfort integration."""

from dataclasses import dataclass, field
from typing import Any

from incomfortclient import Gateway as InComfortGateway, Heater as InComfortHeater

from homeassistant.const import CONF_HOST, CONF_PASSWORD
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.util.hass_dict import HassKey

from .const import DOMAIN


@dataclass
class InComfortData:
    """Keep the Intergas InComfort entry data."""

    client: InComfortGateway
    heaters: list[InComfortHeater] = field(default_factory=list)


DATA_INCOMFORT: HassKey[dict[str, InComfortData]] = HassKey(DOMAIN)


ERROR_STATUS_MAPPING: dict[int, tuple[str, str]] = {
    401: (CONF_PASSWORD, "auth_error"),
    404: ("base", "not_found"),
}


async def async_connect_gateway(
    hass: HomeAssistant,
    entry_data: dict[str, Any],
) -> InComfortData:
    """Validate the configuration."""
    credentials = dict(entry_data)
    hostname = credentials.pop(CONF_HOST)

    client = InComfortGateway(
        hostname, **credentials, session=async_get_clientsession(hass)
    )
    heaters = await client.heaters()

    return InComfortData(client=client, heaters=heaters)
