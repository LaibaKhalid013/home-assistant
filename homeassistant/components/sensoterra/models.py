"""Sensoterra models."""

from datetime import datetime
from enum import StrEnum, auto

from sensoterra.probe import Probe, Sensor

from homeassistant.helpers.update_coordinator import UpdateFailed


class ProbeSensorType(StrEnum):
    """Generic sensors within a Sensoterra probe."""

    MOISTURE = auto()
    SI = auto()
    TEMPERATURE = auto()
    BATTERY = auto()
    RSSI = auto()


class SensoterraSensor:
    """Model Sensoterra probe data produced by the API."""

    depth: int | None = None
    soil: str | None = None
    timestamp: datetime | None = None
    value: float | None = None
    type: ProbeSensorType

    def __init__(
        self,
        probe: Probe,
        sensor: Sensor | ProbeSensorType,
    ) -> None:
        """Initialise Sensoterra sensor."""
        self.name = probe.name
        self.sku = probe.sku
        self.serial = probe.serial
        self.location = probe.location
        if isinstance(sensor, Sensor):
            self.id = sensor.id
            self.type = ProbeSensorType[sensor.type]
            self.depth = sensor.depth
            self.soil = sensor.soil
            self.value = sensor.value
            self.timestamp = sensor.timestamp
        else:
            self.id = f"{probe.serial}-{sensor}"
            self.type = sensor
            if self.type == ProbeSensorType.RSSI:
                if probe.rssi is not None:
                    self.value = probe.rssi
                    self.timestamp = probe.timestamp
            elif self.type == ProbeSensorType.BATTERY:
                if probe.battery is not None:
                    self.value = probe.battery
                    self.timestamp = probe.timestamp
            else:
                raise UpdateFailed(f"Unknown sensor {sensor}")
