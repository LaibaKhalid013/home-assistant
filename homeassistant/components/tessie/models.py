"""The Tessie integration models."""

from __future__ import annotations

from dataclasses import dataclass

from homeassistant.helpers.device_registry import DeviceInfo

from .coordinator import TessieStateUpdateCoordinator


@dataclass
class TessieData:
    """Data for the Tessie integration."""

    vehicles: list[TessieVehicleData]


@dataclass
class TessieVehicleData:
    """Data for a Tessie vehicle."""

    data_coordinator: TessieStateUpdateCoordinator
    device: DeviceInfo
    vin: str
