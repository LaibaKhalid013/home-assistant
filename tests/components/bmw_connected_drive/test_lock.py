"""Test BMW locks."""

from unittest.mock import AsyncMock

from bimmer_connected.models import MyBMWRemoteServiceError
from bimmer_connected.vehicle.remote_services import RemoteServices
from freezegun import freeze_time
import pytest
import respx
from syrupy.assertion import SnapshotAssertion

from homeassistant.core import HomeAssistant
from homeassistant.exceptions import HomeAssistantError

from . import check_remote_service_call, setup_mocked_integration


@freeze_time("2023-06-22 10:30:00+00:00")
async def test_entity_state_attrs(
    hass: HomeAssistant,
    bmw_fixture: respx.Router,
    snapshot: SnapshotAssertion,
) -> None:
    """Test lock states and attributes."""

    # Setup component
    assert await setup_mocked_integration(hass)

    # Get all select entities
    assert hass.states.async_all("lock") == snapshot


@pytest.mark.parametrize(
    ("entity_id", "new_value", "old_value", "service", "remote_service"),
    [
        (
            "lock.m340i_xdrive_lock",
            "locked",
            "unlocked",
            "lock",
            "door-lock",
        ),
        ("lock.m340i_xdrive_lock", "unlocked", "locked", "unlock", "door-unlock"),
    ],
)
async def test_service_call_success(
    hass: HomeAssistant,
    entity_id: str,
    new_value: str,
    old_value: str,
    service: str,
    remote_service: str,
    bmw_fixture: respx.Router,
) -> None:
    """Test successful input change."""

    # Setup component
    assert await setup_mocked_integration(hass)
    hass.states.async_set(entity_id, old_value)
    assert hass.states.get(entity_id).state == old_value

    # Test
    await hass.services.async_call(
        "lock",
        service,
        blocking=True,
        target={"entity_id": entity_id},
    )
    check_remote_service_call(bmw_fixture, remote_service)
    assert hass.states.get(entity_id).state == new_value


@pytest.mark.parametrize(
    ("entity_id", "service"),
    [
        ("lock.m340i_xdrive_lock", "lock"),
        ("lock.m340i_xdrive_lock", "unlock"),
    ],
)
async def test_service_call_fail(
    hass: HomeAssistant,
    entity_id: str,
    service: str,
    bmw_fixture: respx.Router,
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Test failed button press."""

    # Setup component
    assert await setup_mocked_integration(hass)
    old_value = hass.states.get(entity_id).state

    # Setup exception
    monkeypatch.setattr(
        RemoteServices,
        "trigger_remote_service",
        AsyncMock(side_effect=MyBMWRemoteServiceError),
    )

    # Test
    with pytest.raises(HomeAssistantError):
        await hass.services.async_call(
            "lock",
            service,
            blocking=True,
            target={"entity_id": entity_id},
        )
    assert hass.states.get(entity_id).state == old_value
