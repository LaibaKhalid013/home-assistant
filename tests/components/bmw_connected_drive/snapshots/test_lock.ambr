# serializer version: 1
# name: test_entity_state_attrs
  list([
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'attribution': 'Data provided by MyBMW',
        'car': 'iX xDrive50',
        'door_lock_state': 'LOCKED',
        'friendly_name': 'iX xDrive50 Lock',
        'supported_features': <LockEntityFeature: 0>,
        'vin': 'WBA00000000DEMO01',
      }),
      'context': <ANY>,
      'entity_id': 'lock.ix_xdrive50_lock',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'locked',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'attribution': 'Data provided by MyBMW',
        'car': 'i4 eDrive40',
        'door_lock_state': 'LOCKED',
        'friendly_name': 'i4 eDrive40 Lock',
        'supported_features': <LockEntityFeature: 0>,
        'vin': 'WBA00000000DEMO02',
      }),
      'context': <ANY>,
      'entity_id': 'lock.i4_edrive40_lock',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'locked',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'attribution': 'Data provided by MyBMW',
        'car': 'M340i xDrive',
        'door_lock_state': 'LOCKED',
        'friendly_name': 'M340i xDrive Lock',
        'supported_features': <LockEntityFeature: 0>,
        'vin': 'WBA00000000DEMO03',
      }),
      'context': <ANY>,
      'entity_id': 'lock.m340i_xdrive_lock',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'locked',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'attribution': 'Data provided by MyBMW',
        'car': 'i3 (+ REX)',
        'door_lock_state': 'UNLOCKED',
        'friendly_name': 'i3 (+ REX) Lock',
        'supported_features': <LockEntityFeature: 0>,
        'vin': 'WBY00000000REXI01',
      }),
      'context': <ANY>,
      'entity_id': 'lock.i3_rex_lock',
      'last_changed': <ANY>,
      'last_reported': <ANY>,
      'last_updated': <ANY>,
      'state': 'unlocked',
    }),
  ])
# ---
